src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -2170,197 +2159,103 @@ public BigDecimal sqrt(MathContext mc) {
-                MathContext mcTmp = new MathContext(targetPrecision, tmpRm);
-                result = approx.scaleByPowerOfTen(-scaleAdjust/2).round(mcTmp);
-
-                // If result*result != this numerically, the square
-                // root isn't exact
-                if (this.subtract(result.square()).compareTo(ZERO) != 0) {
+                BigInteger[] sqrtRem = stripped.unscaledValue().sqrtAndRemainder();
+                result = new BigDecimal(sqrtRem[0], strippedScale >> 1);
+
+                // If result*result != this numerically or requires too high precision,
+                // the square root isn't exact
+                if (sqrtRem[1].signum != 0 || mc.precision != 0 && result.precision() > mc.precision)
                     throw new ArithmeticException("Computed square root not exact.");
-                }
             } else {
-                result = approx.scaleByPowerOfTen(-scaleAdjust/2).round(mc);
-
-                switch (targetRm) {
-                case DOWN:
-                case FLOOR:
-                    // Check if too big
-                    if (result.square().compareTo(this) > 0) {
-                        BigDecimal ulp = result.ulp();
-                        // Adjust increment down in case of 1.0 = 10^0
-                        // since the next smaller number is only 1/10
-                        // as far way as the next larger at exponent
-                        // boundaries. Test approx and *not* result to
-                        // avoid having to detect an arbitrary power
-                        // of ten.
-                        if (approx.compareTo(ONE) == 0) {
-                            ulp = ulp.multiply(ONE_TENTH);
+                // To allow BigInteger.sqrt() to be used to get the square
+                // root, it is necessary to normalize the input so that
+                // its integer part is sufficient to get the square root
+                // with the desired precision.
+
+                final boolean halfWay = mc.roundingMode.isHalfWay();
+                // To obtain a square root with N digits,
+                // the radicand must have at least 2*(N-1)+1 == 2*N-1 digits.
+                final long minWorkingPrec = ((mc.precision + (halfWay ? 1L : 0L)) << 1) - 1L;
+                // normScale is the number of digits to take from the fraction of the input
+                long normScale = minWorkingPrec - this.precision() + this.scale;
+                normScale += normScale & 1L; // the scale for normalizing must be even
+
+                final long workingScale = this.scale - normScale;
+                if (workingScale != (int) workingScale)
+                    throw new ArithmeticException("Overflow");
+
+                BigDecimal working = new BigDecimal(this.intVal, this.intCompact, (int) workingScale, this.precision);
+                BigInteger workingInt = working.toBigInteger();
```

Comment:
@rgiulietti The only solution I can think of to avoid this is to reimplement Zimmerman's square root algorithm in the class `BigDecimal`, using the radix 10 representation of the `BigDecimal`'s magnitude.

Reply:
After looking at the code more closely, these are _not_ the reasons for the long running times of my example above.
Neither the computation of the large power of 10, nor the multiplication, nor the `BigInteger` square root are slow.

The slowdown in my example of sqrt(121) above comes from the call to `stripZerosToMatchScale()`, which performs about a million divisions by 10 to meet the preferred scale.

In fact, when there are no trailing zeros in the result, the running times are very fast.
Replacing 121 (an exact square) with 122 in the example above reduces the time to less than 1 s rather than more than 3 min, because there are no trailing zeros to get rid of:
`BigDecimal.valueOf(121).sqrt(new MathContext(1_000_000, RoundingMode.HALF_EVEN))` more than 3 min
`BigDecimal.valueOf(122).sqrt(new MathContext(1_000_000, RoundingMode.HALF_EVEN))` less than 1 s

Of course, this is not the fault of `BigDecimal.sqrt()`, but of the way `stripZerosToMatchScale()` is implemented. Maybe the topic of your next PR ;-) ? Might be hard, though.

Reply:
A possible faster implementation of `stripZerosToMatchScale()` could be based on dividing by increasing powers of 10 whose exponent doubles at each step (a kind of repeated squares trick), although the asymptotic running time would still be exponential.

Reply:
Without experimenting a bit, it's hard to tell.
Anyway, I believe the current implementation of `stripZerosToMatchScale()` is O(n^2) rather than exponential, because each single division by 10 is linear.

But let's focus on `sqrt()` in this PR and leave `stripZerosToMatchScale()` for another time.

Reply:
> Without experimenting a bit, it's hard to tell. Anyway, I believe the current implementation of `stripZerosToMatchScale()` is O(n^2) rather than exponential, because each single division by 10 is linear.
> 
> But let's focus on `sqrt()` in this PR and leave `stripZerosToMatchScale()` for another time.

Yes, it is linear, but in the length of the magnitude, while it is exponential in the length of the precision...

Reply:
@rgiulietti  I implemented and tested stripping zeros  algorithm with repeated squares trick, and on my laptop it takes around two seconds for strip one million of zeros. So, I decided to remove the method `stripZerosToEvenScale()`, in order to use only the `stripZerosToMatchScale()` logic.

Reply:
Nice!

Perhaps it would make sense to have your new implementation of `stripZerosToMatchScale()` in another PR, integrate that one first, and then come back to this PR. For general usage of `BigDecimal`, its more useful to have a faster `stripZerosToMatchScale()` than a faster `sqrt()`, I think.

Let me know if you agree of if you prefer to go ahead with with PR first.

Reply:
@fabioromano1 I agree with starting a new PR for speed up `stripZerosToMatchScale()`.

Reply:
I'll open a JBS issue to refer to as soon as you open the new PR.

Please keep the first PR comment as short as possible, then add further details in another comment. As only the first comment is auto-copied in every email in the mailing list, keeping it short helps with bandwidth.

Reply:
Thanks.
To link it to the JBS issue, the title of that PR should read
8341470: BigDecimal.stripTrailingZeros() optimization